### PR TITLE
Remove groupcache

### PR DESCRIPTION
**Pull request by MasslessParticle**
As discussed during the caching retro, this PR removes the experimental distributed cache based on Mailgun's fork of groupcache.

### CODE REVIEW

**Comment by periklis** [comment]
> As discussed during the caching retro, this PR removes the experimental distributed cache based on Mailgun's fork of groupcache.
> 
> 

@MasslessParticle can you share any particular details on what is the rationale for removing groupcache?

**Comment by liguozhong** [comment]
Really bad news, I'm not familiar with c and c++, and when something goes wrong with redis or memcache I can't locate and fix it. 

I really hope to have a cache written by go.

**Comment by MasslessParticle** [comment]
@periklis @liguozhong 

Thanks for your interest and sorry to come off as opaque. 
[Here](https://github.com/grafana/loki/files/9549368/ADR_.In-Process.Distributed.Cache.pdf) is an Architectural Decision Record for the cache work that explains why we're ultimately removing it.

**TL;DR:**
1. Vanilla Groupcache isn't a good fit for Loki's caching because it's purely read-through
2. As a result we used a fork built by the folks at Mailgun that supports `Set` operations
3. Under load, we encountered a deadlock that causes cascading failure across the cluster
4. We fear that the deadlock issue is probabilistic so we deemed it safer to remove the cache altogether and potentially revisit the issue at a later date.

**Comment by acukan** [comment]
Does it mean we are back to memcached or you guys suggest redis on meanwhile?

**Comment by MasslessParticle** [comment]
@acukan We're going to continue using memcached. We have support for Redis but it's not tested in production environments.

File: pkg/loki/config_wrapper_test.go
Comment (dannykopping) [inline pkg/loki/config_wrapper_test.go:833]: Are we sticking with `embedded_cache` as the alias for fifocache?

**Review by dannykopping — COMMENTED** [review]
LGTM, with a question about `embedded_cache`

File: pkg/loki/config_wrapper_test.go
Comment (MasslessParticle) [reply pkg/loki/config_wrapper_test.go:833]: Last I heard, we were

Comment (dannykopping) [reply pkg/loki/config_wrapper_test.go:833]: :+1: OK cool!

**Review by dannykopping — APPROVED** [review]
LGTM, once the conflicts are resolved we can merge

**Review by kavirajk — APPROVED** [review]

**Pull request merged**
